### PR TITLE
[Docs] update the PT 2.0 optimization doc with latest findings

### PR DESCRIPTION
**Pull request by sayakpaul**

### CODE REVIEW

File: docs/source/en/optimization/torch2.0.mdx [hunk 1]
```diff
@@ -60,6 +60,19 @@ pip install --upgrade torch torchvision diffusers
...
 
+    If you want to use the vanilla attention processor (([`~diffusers.models.attention_processor.AttnProcessor`])) as shown below:
+
+    ```Python
+    import torch
+    from diffusers import DiffusionPipeline
+    from diffusers.models.attention_processor import AttnProcessor
+
+    pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16).to("cuda")
+    pipe.unet.set_attn_processor(AttnProcessor())
```
Comment (patrickvonplaten) [inline docs/source/en/optimization/torch2.0.mdx:71]:
```suggestion
    pipe.unet.set_default_attn_processor()
```
does the same.

File: docs/source/en/optimization/torch2.0.mdx [hunk 2]
```diff
@@ -80,131 +93,78 @@ pip install --upgrade torch torchvision diffusers
 
     Depending on the type of GPU, `compile()` can yield between 2-9% of _additional speed-up_ over the accelerated transformer optimizations. Note, however, that compilation is able to squeeze more performance improvements in more recent GPU architectures such as Ampere (A100, 3090), Ada (4090) and Hopper (H100).
     
-    Compilation takes some time to complete, so it is best suited for situations where you need to prepare your pipeline once and then perform the same type of inference operations multiple times.
+    Compilation takes some time to complete, so it is best suited for situations where you need to prepare your pipeline once and then perform the same type of inference operations multiple times. Calling the compiled pipeline on a different type of input will retrigger compilation which can be expensive.
```
Comment (patrickvonplaten) [inline docs/source/en/optimization/torch2.0.mdx:96]:
```suggestion
    Compilation takes some time to complete, so it is best suited for situations where you need to prepare your pipeline once and then perform the same type of inference operations multiple times. Calling the compiled pipeline on a different image size will re-trigger compilation which can be expensive.
```

File: docs/source/en/optimization/torch2.0.mdx [hunk 3]
```diff
@@ -80,131 +93,78 @@ pip install --upgrade torch torchvision diffusers
...
-
-
-(1) Batch Size >= 32 requires enable_vae_slicing() because of https://github.com/pytorch/pytorch/issues/81665.
-This is required for PyTorch 1.13.1, and also for PyTorch 2.0 and large batch sizes.
-
-For more details about how this benchmark was run, please refer to [this PR](https://github.com/huggingface/diffusers/pull/2303) and to [the blog post](https://pytorch.org/blog/accelerated-diffusers-pt-20/).
+We conducted a comprehensive benchmark with PyTorch 2.0's efficient attention implementation and `torch.compile` across different GPUs and different batch sizes for five of our most used pipelines. 
+In the following tables, we report our findings in terms of the number of iterations processed per second. 
+
+### A100 (batch size: 1)
```
Comment (patrickvonplaten) [inline docs/source/en/optimization/torch2.0.mdx:104]: Let's also add RTX4090 and T4 no? 

Think people will be quite interested in the home GPUs here

File: docs/source/en/optimization/torch2.0.mdx [hunk 4]
```diff
@@ -80,131 +93,78 @@ pip install --upgrade torch torchvision diffusers
 
     Depending on the type of GPU, `compile()` can yield between 2-9% of _additional speed-up_ over the accelerated transformer optimizations. Note, however, that compilation is able to squeeze more performance improvements in more recent GPU architectures such as Ampere (A100, 3090), Ada (4090) and Hopper (H100).
```
Comment (patrickvonplaten) [inline docs/source/en/optimization/torch2.0.mdx:95]: Actually can we change the code here? I would maybe structure it as follows:

1. Explain how torch compile can be used for every pipeline out there by compiling the unet:
=> Just show:
```py
pipe.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)
```

2. Then give examples for the most important models:
- Stable Diffusion
- ControlNet
- IF
=> here we can copy paste our benchmark code

File: docs/source/en/optimization/torch2.0.mdx (hunk 3 above)
Comment (sayakpaul) [reply docs/source/en/optimization/torch2.0.mdx:104]: There is a reason why the PR is in the draft mode :-)

File: docs/source/en/optimization/torch2.0.mdx [hunk 5]
```diff
@@ -12,19 +12,19 @@ specific language governing permissions and limitations under the License.
...
 ## Installation
-To benefit from the accelerated attention implementation and `torch.compile`, you just need to install the latest versions of PyTorch 2.0 from `pip`, and make sure you are on diffusers 0.13.0 or later. As explained below, `diffusers` automatically uses the attention optimizations (but not `torch.compile`) when available.
+To benefit from the accelerated attention implementation and `torch.compile`, you just need to install the latest versions of PyTorch 2.0 from `pip`, and make sure you are on diffusers 0.13.0 or later. As explained below, `diffusers` automatically uses the optimized attention processor ([`~diffusers.models.attention_processor.AttnProcessor2_0`]) (but not `torch.compile`) when PyTorch 2.0 is available.
 
 ```bash
 pip install --upgrade torch torchvision diffusers
 ```
 
-## Using accelerated transformers and torch.compile.
+## Using accelerated transformers and `torch.compile`.
```
Comment (pcuenca) [inline docs/source/en/optimization/torch2.0.mdx:27]: I _think_ the backquotes were not properly shown in headings when docs were generated? Let's just keep an eye on it and remove them if we see anything weird after merging :)

File: docs/source/en/optimization/torch2.0.mdx [hunk 6]
```diff
@@ -46,165 +46,381 @@ pip install --upgrade torch torchvision diffusers
...
-    pipe.unet.set_attn_processor(AttnProcessor2_0())
+    + pipe.unet.set_attn_processor(AttnProcessor2_0())
 
     prompt = "a photo of an astronaut riding a horse on mars"
     image = pipe(prompt).images[0]
     ```
 
     This should be as fast and memory efficient as `xFormers`. More details [in our benchmark](#benchmark).
 
+    If you want to revert to use the vanilla attention processor ([`~AttnProcessor`]) which can help to make the pipeline more deterministic, you can use the [`~UNet2DModel.set_default_attn_processor`] function:
```
Comment (pcuenca) [inline docs/source/en/optimization/torch2.0.mdx:63]:
```suggestion
    It is possible to revert to the vanilla attention processor ([`~AttnProcessor`]), which can be helpful to make the pipeline more deterministic, or if you need to convert a fine-tuned model to other formats such as [Core ML](https://huggingface.co/docs/diffusers/v0.16.0/en/optimization/coreml#how-to-run-stable-diffusion-with-core-ml). To use the normal attention processor you can use the [`~UNet2DModel.set_default_attn_processor`] function:
```

File: docs/source/en/optimization/torch2.0.mdx [hunk 7]
```diff
@@ -46,165 +46,381 @@ pip install --upgrade torch torchvision diffusers
...
-| 4090 | 16 |  | 38.62 | 35.65 | 34.07 | 11.78 |  |
-| 4090 | 32 (1) |  | 76.57 | 69.48 | 65.35 | 14.65 |  |
-| 4090 | 48 |  | 114.44 | 106.3 |  | 7.11 |  |
-
-
-(1) Batch Size >= 32 requires enable_vae_slicing() because of https://github.com/pytorch/pytorch/issues/81665.
-This is required for PyTorch 1.13.1, and also for PyTorch 2.0 and large batch sizes.
-
-For more details about how this benchmark was run, please refer to [this PR](https://github.com/huggingface/diffusers/pull/2303) and to [the blog post](https://pytorch.org/blog/accelerated-diffusers-pt-20/).
+We conducted a comprehensive benchmark with PyTorch 2.0's efficient attention implementation and `torch.compile` across different GPUs and different batch sizes for five of our most used pipelines.
```
Comment (pcuenca) [inline docs/source/en/optimization/torch2.0.mdx:94]:
```suggestion
We conducted a comprehensive benchmark with PyTorch 2.0's efficient attention implementation and `torch.compile` across different GPUs and batch sizes for five of our most used pipelines. We used `diffusers 0.17.0.dev0`, which [makes sure `torch.compile()` is leveraged optimally](https://github.com/huggingface/diffusers/pull/3286).
```

I think it's important to state the diffusers version here, as the beginning of the document only mentions `0.13.0` and it wouldn't be possible to achieve these numbers with it. For complete reproducibility disclosure, I'd state the revision number if we have it (I think it's in your Docker images, right @sayakpaul?).

**Review by pcuenca — COMMENTED** [review]
Awesome! Do you want me to regenerate some of the plots we used for [the blog post](https://pytorch.org/blog/accelerated-diffusers-pt-20/)? Let me know if you'd like to use any of them here and I'll prepare them.

**Comment by sayakpaul** [comment]
> Do you want me to regenerate some of the plots we used for [the blog post](https://pytorch.org/blog/accelerated-diffusers-pt-20/)?

Sure, that'd be very helpful!

**Review by patrickvonplaten — APPROVED** [review]
Love it :heart_eyes: Super clean and nicely written. 
I'm sure the community will love this! 

Some graphs could definitely help (maybe we could use those as well to update the official PyTorch blog post cc @pcuenca )

**Comment by sayakpaul** [comment]
For the failing test: https://github.com/huggingface/diffusers/pull/3397#issuecomment-1545357076

File: docs/source/en/optimization/torch2.0.mdx [hunk 8]
```diff
@@ -46,165 +47,397 @@ pip install --upgrade torch torchvision diffusers
...
+    [torch compile docs](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html).
+
+    ```python
+    pipe.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)
     images = pipe(prompt, num_inference_steps=steps, num_images_per_prompt=batch_size).images
     ```
 
-    Depending on the type of GPU, `compile()` can yield between 2-9% of _additional speed-up_ over the accelerated transformer optimizations. Note, however, that compilation is able to squeeze more performance improvements in more recent GPU architectures such as Ampere (A100, 3090), Ada (4090) and Hopper (H100).
+    Depending on the type of GPU, `compile()` can yield between **3% - 56%** of _additional speed-up_ over the accelerated transformer optimizations. Note, however, that compilation is able to squeeze more performance improvements in more recent GPU architectures such as Ampere (A100, 3090), Ada (4090) and Hopper (H100).
```
Comment (pcuenca) [inline docs/source/en/optimization/torch2.0.mdx:88]: It's actually much more than that, especially in the case of IF. I _think_ the percent computation was wrong in the excel, so I changed it. For example, in A100 (BS 1), looking at txt2img in the stable version it goes from 21.66 it/s to 44.03 it/s, which is a bit more than double. Hence, the percent improvement is 103. Can you please double check @sayakpaul?

We could for example say that we get up to twice as many iterations per second, or almost 5x in the case of IF Stage I.

I'm also curious and surprised that the improvement is so big (especially in IF)! Do you have any insight on that?

File: docs/source/en/optimization/torch2.0.mdx [hunk 9]
```diff
@@ -266,6 +267,22 @@ for _ in range(3):
...
+
+![t2i_speedup](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/pt2_benchmarks/t2i_speedup.png)
+
+To give you an even better idea of how this speed-up holds for the other pipelines presented above, consider the following 
+plot that shows the benchmarking numbers from an A100 across three different batch sizes
+(with PyTorch 2.0 nightly and `torch.compile()`):
+
+![a100_numbers](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/pt2_benchmarks/a100_numbers.png)
+
+_(Our benchmarking metric for the plots above is **number of iterations/second**)_
```
Comment (pcuenca) [inline docs/source/en/optimization/torch2.0.mdx:282]: Good call!

**Review by pcuenca — APPROVED** [review]
Amazing, thanks a lot @sayakpaul!

File: docs/source/en/optimization/torch2.0.mdx (hunk 8 above)
Comment (sayakpaul) [reply docs/source/en/optimization/torch2.0.mdx:88]: I did `100 * (b - a) / b` following https://docs.google.com/spreadsheets/d/1LrltKSgZyOZiLQ7n7_GvIl_BoED-AHeVOJYBc_QqxXA/edit#gid=0

> I'm also curious and surprised that the improvement is so big (especially in IF)! Do you have any insight on that?

It's probably just better suited for tiling and fusion and collectively they might contribute in improving the overall arithmetic density while minimizing the memory transfers. But need to profile to say for sure.

**Comment by pcuenca** [comment]
We may want to update this section in a follow up PR: https://huggingface.co/docs/diffusers/stable_diffusion#next-steps

**Comment by patrickvonplaten** [comment]
Cool should we merge this one?

**Comment by sayakpaul** [comment]
Waiting for @pcuenca to clarify:

https://github.com/huggingface/diffusers/pull/3370#discussion_r1192057800

File: docs/source/en/optimization/torch2.0.mdx (hunk 8 above)
Comment (sayakpaul) [reply docs/source/en/optimization/torch2.0.mdx:88]: You're actually right. Let me edit and merge.

**Pull request merged**